### PR TITLE
[REVERTED] GroupedIterator improvements

### PR DESCRIPTION
**Pull request by joroKr21**
- Specify the size when creating `ArrayBuffer`s
 - Modify intermediate results in place as much as possible
 - Don't allocate an `Option` for padding

### CODE REVIEW

**Comment by SethTisue** [comment]
@scala/collections batsignal

File: src/library/scala/collection/Iterator.scala [hunk 1]
```diff
@@ -212,44 +211,39 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
       buf
     }
 
-    private def padding(x: Int) = immutable.ArraySeq.untagged.fill(x)(pad.get())
-    private def gap = (step - size) max 0
+    private def gap =
+      (step - size) max 0
```
Comment (NthPortal) [inline src/library/scala/collection/Iterator.scala]: why is this on a new line?

File: src/library/scala/collection/Iterator.scala [hunk 2]
```diff
@@ -212,44 +211,39 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
...
-      val xs: Seq[B] = {
-        val res = takeDestructively(count)
-        // was: extra checks so we don't calculate length unless there's reason
-        // but since we took the group eagerly, just use the fast length
-        val shortBy = count - res.length
-        if (shortBy > 0 && pad.isDefined) res ++ padding(shortBy) else res
+      val xs = {
+        // If there is padding defined we insert it immediately
+        // so the rest of the code can be oblivious
+        val xs = takeDestructively(count)
```
Comment (NthPortal) [inline src/library/scala/collection/Iterator.scala]: I do not like defining `xs` inside of `xs`. it's too easy to get confused

File: src/library/scala/collection/Iterator.scala [hunk 3]
```diff
@@ -262,14 +256,13 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
       else go(step)
     }
 
-    def hasNext = filled || fill()
+    def hasNext: Boolean =
+      filled || fill()
```
Comment (NthPortal) [inline src/library/scala/collection/Iterator.scala]: again, not sure why this is on a new line. you're doing a whole bunch of reducing the number of lines in other places, but then add new lines here. this is a nitpick—I'm just confused

File: src/library/scala/collection/Iterator.scala [hunk 4]
```diff
@@ -262,14 +256,13 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
...
+
     @throws[NoSuchElementException]
     def next(): immutable.Seq[B] = {
-      if (!filled)
-        fill()
-
-      if (!filled)
-        throw new NoSuchElementException("next on empty iterator")
+      if (!filled) fill()
+      if (!filled) throw new NoSuchElementException("next on empty iterator")
```
Comment (NthPortal) [inline src/library/scala/collection/Iterator.scala]: if we're already editing this, `Iterator.empty.next()` would be the idiomatic way of doing this (it obviates the need to keep the exception message consistent across lots of implementations)

**Review by NthPortal — APPROVED** [review]
LGTM. I have a few nitpicks, but none of them are blockers

File: src/library/scala/collection/Iterator.scala (hunk 1 above)
Comment (joroKr21) [reply src/library/scala/collection/Iterator.scala]: No good reason, just preference - I can put it back on one line

File: src/library/scala/collection/Iterator.scala (hunk 2 above)
Comment (joroKr21) [reply src/library/scala/collection/Iterator.scala]: Actually I don't think we need this nested block

File: src/library/scala/collection/Iterator.scala (hunk 1 above)
Comment (som-snytt) [reply src/library/scala/collection/Iterator.scala]: because it adds a "gap".

File: src/library/scala/collection/Iterator.scala (hunk 3 above)
Comment (som-snytt) [reply src/library/scala/collection/Iterator.scala]: because the method "hasNext" line.

File: src/library/scala/collection/Iterator.scala (hunk 4 above)
Comment (som-snytt) [reply src/library/scala/collection/Iterator.scala]: it's a one-liner so the line is filled.

**Comment by joroKr21** [comment]
Btw I also noticed some bound checks before calling `take` or `dropInPlace` / similar. Does it make sense to remove them? `take` and `drop` variants work with any integer and do what you would expect.

File: src/library/scala/collection/Iterator.scala [hunk 5]
```diff
@@ -146,19 +146,18 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
...
    */
   class GroupedIterator[B >: A](self: Iterator[B], size: Int, step: Int) extends AbstractIterator[immutable.Seq[B]] {
-
     require(size >= 1 && step >= 1, f"size=$size%d and step=$step%d, but both must be positive")
 
-    private[this] var buffer: ArrayBuffer[B] = ArrayBuffer()  // the buffer
-    private[this] var filled = false                          // whether the buffer is "hot"
-    private[this] var _partial = true                         // whether we deliver short sequences
-    private[this] var pad: Option[() => B] = None             // what to pad short sequences with
+    private[this] val group = new ArrayBuffer[B](size)  // the group
```
Comment (som-snytt) [inline src/library/scala/collection/Iterator.scala:157]: TIL it doesn't lint `private[this]` with no accessors.
```
private var buffer in class GroupedIterator is never updated: consider using immutable val
```

**Comment by som-snytt** [comment]
There are so many min checks. I wonder if the "NB" on `takeDestructively` is still correct. I only remember it took me a while to understand enough to answer "For reasons which remain to be determined". Also not sure if the other comment was still relevant, which pertained to the half-line fix.

File: src/library/scala/collection/Iterator.scala [hunk 6]
```diff
@@ -212,66 +211,55 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
...
-      val prevSize = buffer.size
+      val prevSize = group.size
       def isFirst = prevSize == 0
+      val extension = takeDestructively(count)
       // If there is padding defined we insert it immediately
       // so the rest of the code can be oblivious
-      val xs: Seq[B] = {
-        val res = takeDestructively(count)
-        // was: extra checks so we don't calculate length unless there's reason
-        // but since we took the group eagerly, just use the fast length
```
Comment (som-snytt) [inline src/library/scala/collection/Iterator.scala:226]: This was the other comment from the previous fix to padding. The strawman merge makes history hard to follow up, so I picked it up on 2.12.

File: src/library/scala/collection/Iterator.scala [hunk 7]
```diff
@@ -212,66 +211,55 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
...
-        val shortBy = count - res.length
-        if (shortBy > 0 && pad.isDefined) res ++ padding(shortBy) else res
+      var shortBy = count - extension.size
+      if (pad != null) while (shortBy > 0) {
```
Comment (som-snytt) [inline src/library/scala/collection/Iterator.scala:223]: This is probably better than extra `++` (I wonder if these were iterators in an early version of the code), but I wonder no one asked about this indentation. I wonder if jit would optimize `while (pad != null && shortBy > 0)`? Not recommending. But for the amount of time it takes to grok this class, an extra NL is such a luxury.

Comment (som-snytt) [reply src/library/scala/collection/Iterator.scala:223]: Also surprised old paulp expression was not `pad.foreach`.

**Review by som-snytt — APPROVED** [review]

File: src/library/scala/collection/Iterator.scala (hunk 5 above)
Comment (som-snytt) [reply src/library/scala/collection/Iterator.scala:157]: The lint is turned on, so removing `[this]` does make it warn about `var`...

Comment (som-snytt) [reply src/library/scala/collection/Iterator.scala:157]: I created a ticket in case  it is fixable.

**Comment by SethTisue** [comment]
Thanks all!

**Comment by SethTisue** [comment]
reverted in #9859, because of the community build failures detailed at https://github.com/scala/community-build/pull/1519

oh well, win some, lose some! happy to consider a revised version for 2.13.9

**Comment by som-snytt** [comment]
The test is bad, since it is the same as (the small number on my small VM)
```
Welcome to Scala 2.13.7 (OpenJDK 64-Bit Server VM, Java 17).
Type in expressions for evaluation. Or try :help.

scala> Iterator.continually(42).grouped(Int.MaxValue/32)
val res0: Iterator[Int]#GroupedIterator[Int] = <iterator>

scala> .next()
java.lang.OutOfMemoryError: Java heap space
  at scala.collection.mutable.ArrayBuffer$.scala$collection$mutable$ArrayBuffer$$ensureSize(ArrayBuffer.scala:300)
  at scala.collection.mutable.ArrayBuffer.ensureSize(ArrayBuffer.scala:69)
  at scala.collection.mutable.ArrayBuffer.addOne(ArrayBuffer.scala:139)
  at scala.collection.mutable.ArrayBuffer.addOne(ArrayBuffer.scala:43)
  at scala.collection.Iterator$GroupedIterator.takeDestructively(Iterator.scala:209)
  at scala.collection.Iterator$GroupedIterator.go(Iterator.scala:224)
  at scala.collection.Iterator$GroupedIterator.fill(Iterator.scala:261)
  at scala.collection.Iterator$GroupedIterator.next(Iterator.scala:269)
  ... 24 elided
```
I do notice in the PR that on the first group, `group` is allocated and then `extension` is allocated, filled, and then copied over instead of just used.

Probably Nth's fix for ArrayBuffer sizing will help avoid OOM or resizings.
https://github.com/scala/scala/pull/9819

Possibly the possibly huge group size should be allocated only if the iterator has a known size that can be checked.

**Comment by som-snytt** [comment]
Better example, where they ask for it
```
scala> Iterator.continually(42).take(1024).grouped(Int.MaxValue/32).withPadding(-1).next()
java.lang.OutOfMemoryError: Java heap space
  at scala.reflect.ManifestFactory$AnyManifest.newArray(Manifest.scala:328)
  at scala.reflect.ManifestFactory$AnyManifest.newArray(Manifest.scala:327)
  at scala.collection.immutable.ArraySeq$.fill(ArraySeq.scala:291)
  at scala.collection.immutable.ArraySeq$.fill(ArraySeq.scala:273)
  at scala.collection.ClassTagIterableFactory$AnyIterableDelegate.fill(Factory.scala:678)
  at scala.collection.Iterator$GroupedIterator.padding(Iterator.scala:215)
  at scala.collection.Iterator$GroupedIterator.go(Iterator.scala:228)
  at scala.collection.Iterator$GroupedIterator.fill(Iterator.scala:261)
  at scala.collection.Iterator$GroupedIterator.next(Iterator.scala:269)
  ... 23 elided
```
Looking at the API, the client must not give a group size that exhausts memory, because either there is only one group and why are you calling grouped, or it must fail.

**Comment by joroKr21** [comment]
Huh, interesting 🤔  - I will think about it some more. Maybe we can just reuse the initial buffer to reduce the allocation cost and then letting it grow is fine because eventually it will reach a fixed size. But yeah, this can wait for 2.13.9

**Comment by som-snytt** [comment]
I'm taking a swing in the linked PR, switch-hitting.
